src/storage/src/source/reclock.rs

```diff
@@ -7,931 +7,646 @@
+    where
+        IntoTime: TotalOrder,
+    {
+        let mut memo: Option<(FromTime, Result<IntoTime, ReclockError<FromTime>>)> = None;
+        batch.into_iter().map(move |(msg, src_ts)| {
+            let result = match &memo {
+                Some((prev_src_ts, result)) if prev_src_ts == &src_ts => result.clone(),
+                _ => {
+                    let result = self.reclock_time_total(&src_ts);
+                    memo.insert((src_ts, result)).1.clone()
                 }
-            }
-            if let Some((_msg, offset)) = messages.last() {
-                batch_upper.insert(pid, *offset + 1);
-            }
+            };
+            (msg, result)
+        })
+    }
+
+    /// Reclocks a single `FromTime` timestamp into the `IntoTime` time domain.
+    fn reclock_time(
+        &self,
+        src_ts: &FromTime,
+    ) -> Result<Antichain<IntoTime>, ReclockError<FromTime>> {
+        let inner = self.inner.borrow();
+        if !PartialOrder::less_than(&inner.as_of, &inner.upper) {
+            return Err(ReclockError::Uninitialized);
+        }
+        if inner.source_upper.less_equal(src_ts) {
+            return Err(ReclockError::BeyondUpper(src_ts.clone()));
+        }
+        if !inner.source_since.less_equal(src_ts) {
+            return Err(ReclockError::NotBeyondSince(src_ts.clone()));
         }
 
-        // Ensure we have enough bindings
-        for (pid, offset) in batch_upper {
-            let bindings_upper = inner.source_upper.get(pid);
-            if let Some(bindings_upper) = bindings_upper {
-                if &offset > bindings_upper {
-                    trace!("offset {} >= bindings_upper {}", offset, bindings_upper);
-                    return Ok(None);
+        // In order to understand the logic of the following section let's first consider an
+        // example of trying to reclock the FromTime D from this partial ordering:
+        //
+        //     ,--B----D
+        //    /              ,-------F----.
+        //   A              /              \
+        //    `---C--------E---------G------H
+        //
+        // ..into a target time domain where the remap collection varies according to this IntoTime
+        // partial ordering:
+        //
+        //   *----*--.---------------*------*
+        //   t0   t1  \              t2     t3
+        //             `--------------------*
+        //                                  t4
+        // ..and the FromTime frontiers at times t0, t1, t2, t3, t4 accumulate to:
+        //
+        // t0: Antichain{A}
+        // t1: Antichain{B, C}
+        // t2: Antichain{F, G}
+        // t3: Antichain{H}
+        // t4: Antichain{H}
+        //
+        // In the example above the correct answer is {t2, t4}, because this is the smallest
+        // antichain of IntoTime times such that the remap collection accumulates at each one of
+        // them to a FromTime frontier that is not beyond D.
```

Comment:
Shouldn't this be "to a FromTime frontier `f` such that `D` is not beyond `f`", so the other way around? I think you might have the same thing somewhere else in this big comment.

Reply:
Good catch, yes!